### PR TITLE
Refactor Table styles and props with updated examples and tests.

### PR DESCRIPTION
**Pull request by ogupte**
## PR Checklist

- Manually tested across supported browsers
  - [x] Chrome
  - [x] Firefox
  - [x] Safari
  - [x] IE11 (Win 7)
  - [x] Edge (Win 10)
- [x] Unit tests written (`common` at minimum)
- [x] PR has one of the `semver-` labels
- [x] Two core team engineer approvals
- [x] One core team UX approval

# Breaking Changes:

## Table
 - removed prop `hasExtraWhiteSpace: bool` and replaced it with `density: oneOf(['compressed', 'extended'])` defaulting to `extended`.
 - changed prop `hasNoBorder` to `hasBorder` with an opposite boolean value
 - A lot of style changes for Table, see diff https://github.com/appnexus/lucid/pull/231/files#diff-d9e45e4cfffeaed1b27a59a59072ccaa

### Table.Thead
 - change className from `lucid-Table-thead` to `lucid-Table-Thead`

### Table.Tbody
 - change className from `lucid-Table-tbody` to `lucid-Table-Tbody`

### Table.Tr
 - change className from `lucid-Table-row` to `lucid-Table-Tr`
 - removed prop `isHeader`
 - renamed prop `hasDetails` to `isActionable`

### Table.Th
 - change className from `lucid-Table-cell` to `lucid-Table-Th`
 - removed prop `hasCheckbox`
 - removed prop `hasButton`
 - removed prop `hasIcon`
 - updated the rendered output to have a more flex-driven layout to handle sorting icons and resizers

### Table.Td
 - change className from `lucid-Table-cell` to `lucid-Table-Td`
 - removed prop `isAfterRowSpan`
 - removed prop `hasCheckbox`
 - removed prop `hasButton`
 - removed prop `hasIcon`

## ScrollTable
 - renamed prop `isNowrap` to `hasWordWrap`

## DataTable
 - removed prop `hasExtraWhiteSpace`
 - renamed props `hasDetails` to `isActionable`

### CODE REVIEW

File: src/components/Table/Table.less [hunk 1]
```diff
@@ -1,371 +1,299 @@
 @Table-row-height: @size-standard-height - 2;
```
Comment (ogupte) [inline src/components/Table/Table.less]: these less variables need to be cleaned up, there are some unused vars defined here.

File: src/components/Table/Table.less [hunk 2]
```diff
@@ -1,371 +1,299 @@
...
+	position: relative;
+	&::before {
+		position: absolute;
+		left: 0;
+		top: 0;
+		height: 100%;
+		width: @width;
+		background-color: @color-primary;
+		content: '';
+		//opacity: @scale;
```
Comment (ogupte) [inline src/components/Table/Table.less]: remove this

File: src/components/DataTable/DataTable.jsx [hunk 3]
```diff
@@ -49,17 +49,13 @@ const DataTable = createClass({
...
-		hasExtraWhitespace: bool,
-		/**
 		 * Render a checkbox in the first column allowing `onSelect` and `onSelectAll` to be triggered.
 		 */
 		isSelectable: bool,
 		/**
 		 * Render each row item to be navigable, allowing `onRowClick` to be triggered.
 		 */
-		hasDetails: bool,
+		isActionable: bool,
```
Comment (jondlm) [inline src/components/DataTable/DataTable.jsx]: This is a better name!

**Comment by jondlm** [comment]
Might want to do an `ag` for `hasExtraWhitespace`. Looks like there are some left over in examples.

File: src/components/Table/Table.jsx [hunk 4]
```diff
@@ -511,41 +479,194 @@ const Table = createClass({
...
 		);
 	}
 });
 
+/**
+ * mapToGrid
+ *
+ * Returns a 2 dimensional array of cell elements of the given component type. The map function can modify value of a cell.
+ */
+const mapToGrid = function mapToGrid(trList, cellType='td', mapFn=_.property('element')) {
```
Comment (jondlm) [inline src/components/Table/Table.jsx]: Is there a reason this is a `const` and not just a regular function?

Comment (ogupte) [inline src/components/Table/Table.jsx]: whoops looks like a fluke, i'll fix it

File: src/components/ScrollTable/examples/1.standard.jsx [hunk 5]
```diff
@@ -34,140 +34,140 @@ export default React.createClass({
 				<Tbody>
 					<Tr>
 						<Td rowSpan={14} hasBorderRight>RS</Td>
-						<Td hasCheckbox><Checkbox/></Td>
```
Comment (sodiumjoe) [inline src/components/ScrollTable/examples/1.standard.jsx]: nice

File: src/components/ScrollTable/examples/1.standard.jsx [hunk 6]
```diff
@@ -34,140 +34,140 @@ export default React.createClass({
...
+						<Td><Checkbox/></Td>
 						<Td>Text</Td>
-						<Td hasIcon hasBorderLeft hasBorderRight><SuccessIcon/></Td>
-						<Td hasButton><Button size='small'>button</Button></Td>
+						<Td  hasBorderLeft hasBorderRight><SuccessIcon/></Td>
```
Comment (sodiumjoe) [inline src/components/ScrollTable/examples/1.standard.jsx]: extra whitespace

File: src/components/Table/Table.jsx [hunk 7]
```diff
@@ -511,41 +479,194 @@ const Table = createClass({
 		className: string,
 
 		/**
-		 * Adjusts the style of the table to have more spacing within the table cells
+		 * Adjusts the row density of the table to have more or less spacing.
+		 */
+		density: oneOf(['compressed', 'extended']),
```
Comment (sodiumjoe) [inline src/components/Table/Table.jsx]: constants off of statics?

File: src/components/Table/Table.jsx [hunk 8]
```diff
@@ -511,41 +479,194 @@ const Table = createClass({
...
+	const cellRowList = _.map(trList, (trElement) => _.map(filterTypes(trElement.props.children, cellType)));
+	const grid = [];
+
+	if (_.isEmpty(cellRowList)) {
+		return [];
+	}
+
+
+	// iterate over each row
+	for (let rowIndex=0; rowIndex<cellRowList.length; rowIndex++) {
```
Comment (sodiumjoe) [inline src/components/Table/Table.jsx]: any reason not to use `_.each`?

Comment (sodiumjoe) [inline src/components/Table/Table.jsx]: or even better, `_.reduce`

File: src/styles/variables.less [hunk 9]
```diff
@@ -82,7 +82,7 @@
 @border-standardBorder: 1px solid @color-borderColor;
 @border-lightBorder: 1px solid @color-borderColorLight;
 
-@border-table: 1px solid @color-gray;
+@border-table: 1px solid #767b7f;
```
Comment (sodiumjoe) [inline src/styles/variables.less]: why'd this change and should it be a variable

File: src/components/Table/Table.jsx [hunk 10]
```diff
@@ -511,41 +479,194 @@ const Table = createClass({
...
+const mapToGrid = function mapToGrid(trList, cellType='td', mapFn=_.property('element')) {
+	const cellRowList = _.map(trList, (trElement) => _.map(filterTypes(trElement.props.children, cellType)));
+	const grid = [];
+
+	if (_.isEmpty(cellRowList)) {
+		return [];
+	}
+
+
+	// iterate over each row
```
Comment (sodiumjoe) [inline src/components/Table/Table.jsx]: cool comment

File: src/components/Table/Table.less [hunk 11]
```diff
@@ -1,371 +1,299 @@
 @Table-row-height: @size-standard-height - 2;
 @Table-row-heightLarge: @Table-row-height + @size-standard;
 @Table-cell-padding: @size-S;
+@Table-cell-border-color: #e3e3e3;
```
Comment (jondlm) [inline src/components/Table/Table.less]: This should be `@color-gray`

File: src/components/Table/Table.less [hunk 12]
```diff
@@ -1,371 +1,299 @@
...
+
+	// Table header styles
+	& > &-Thead {
+		background-color: @Table-header-backgroundColor;
+		color: @color-white;
+
+		// Table header cell styles
+		& > .lucid-Table-Tr > .lucid-Table-Th {
+			border: @Table-header-cell-border;
+			height: 27px; // effectively min height
```
Comment (jondlm) [inline src/components/Table/Table.less]: `@size-standard-height - 1`

File: src/components/Table/Table.less [hunk 13]
```diff
@@ -1,371 +1,299 @@
...
-.Table-row-hover(@bgcolor, @hasDetails: false) {
-	&:hover {
-		background-color: @bgcolor;
-		.lucid-Table-cell when (@hasDetails = true){
-			&:first-of-type, &.lucid-Table-is-after-rowspan {
-				.hardstop-gradient(to right, @color-primary, @bgcolor, @Table-hover-borderWidth);
-			}
-		}
+// Place a styled box on the left edge of the current element with the given scale and width
+.renderLeftMarker(@scale:1; @width: 4px) {
```
Comment (jondlm) [inline src/components/Table/Table.less]: This will be useful for #233

File: src/components/Table/Table.spec.jsx [hunk 14]
```diff
@@ -297,25 +240,23 @@ describe('Table', () => {
...
 				});
 
 				describe('sortDirection', () => {
 					it('should pass thru to the CaretIcon when `isSorted` is also true', () => {
 						const wrapper = shallow(
-							<Th isSorted sortDirection='up' />
+							<Th isSorted sortDirection='down' />
 						);
+						const containerWrapper = wrapper.find('.lucid-Table-Th-inner-caret');
+						console.log(containerWrapper.debug());
```
Comment (jondlm) [inline src/components/Table/Table.spec.jsx]: Might want to remove these, do a quick `ag console.log`

File: src/styles/variables.less (hunk 9 above)
Comment (jondlm) [inline src/styles/variables.less]: Yeah, it's also [repeated here](https://github.com/appnexus/lucid/pull/231/files#diff-d9e45e4cfffeaed1b27a59a59072ccaaR5)

File: src/components/Table/Table.jsx (hunk 7 above)
Comment (jondlm) [inline src/components/Table/Table.jsx]: The only downside to doing that is the docs don't properly show the string values.

File: src/components/Table/Table.jsx (hunk 8 above)
Comment (ogupte) [inline src/components/Table/Table.jsx]: it was initially using both forEach and reduce but as i debugged issues with the first algorithm, it was just so much easier to debug this way, so i kept it.

File: src/components/Table/Table.jsx (hunk 10 above)
Comment (ogupte) [inline src/components/Table/Table.jsx]: i planned to add line comments throughout this function but i think i was distracted by something else.

File: src/components/Table/Table.spec.jsx (hunk 14 above)
Comment (ogupte) [inline src/components/Table/Table.spec.jsx]: doh!

File: src/styles/variables.less (hunk 9 above)
Comment (ogupte) [inline src/styles/variables.less]: the border color changed to match the header background color

File: src/components/Table/Table.jsx (hunk 7 above)
Comment (sodiumjoe) [inline src/components/Table/Table.jsx]: gross

**Pull request merged**